engine/class_modules/priest/sc_priest.cpp

```diff
@@ -2198,21 +2198,30 @@ int priest_t::shadow_weaving_active_dots( const player_t* target ) const
       const dot_t* vt  = td->dots.vampiric_touch;
       const dot_t* dp  = td->dots.devouring_plague;
 
-      dots = swp->is_ticking() + vt->is_ticking() + dp->is_ticking();
+      // You get mastery benefit for a DoT as if it was active, if you are actively putting that DoT up
+      // So to get the mastery benefit you either have that DoT ticking, or you are casting it
+      const spell_data_t* swp_spell = find_class_spell( "Shadow Word: Pain" );
+      const spell_data_t* vt_spell = find_class_spell( "Vampiric Touch" );
+      const spell_data_t* dp_spell = find_class_spell( "Devouring Plague" );
```

Comment:
This works, but there is probably a more elegant way. This is highly inefficient, as you're performing 3x 'find_class_spell' on every damage instance that benefits from Mastery.

Instead of sending the ID, maybe send the whole action pointer and do a class check if it is shadow_word_pain_t, vampiric_touch_t, or devouring_plague_t ?

Reply:
using the action id is fine, as long as the lookup of spell data is moved out of here.

You just need to decide where the right place is to put those. Either just put it into priest.specs, or add a new struct.

Reply:
is there a way to pull it out of the `dot_t` object? I tried doing that but it didnt look like it populated correctly.

Reply:
> is there a way to pull it out of the `dot_t` object? I tried doing that but it didnt look like it populated correctly. Since we already do `const dot_t* swp = td->dots.shadow_word_pain`

dots.shadow_word_pain->current_action->internal_id and pass down the action->internal_id perhaps? But should be a fair amount of ways to use the current_action for comparisons.